### PR TITLE
feat: improve audience handling

### LINKED ISSUE #3747: Token Refactor Step 5: Improve Audience Handling
# Feature Request

Currently, the `aud` claim of tokens is always derived from the remote message's `counterPartyAddress` property, which contains the DSP callback address, which 
is obviously incorrect.

What the `aud` claim is supposed to contain will depend on the type of `IdentityService`: IATP may use a DID, OAuth2 may use a URL, etc. To accommodate that, the `RemoteMessage` will 
get a `counterPartyId` field which is assigned by the `IdentityService` implementation upon message egress and can be obtained from a prior `ContractNegotiation` or `TransferProcess`.

> On message ingress, the `IdentityService` validates the audience. This is already in place and does not need to change.

## Which Areas Would Be Affected?

dsp protocol layers

## Why Is the Feature Desired?

Use the correct audience depending on the `IdentityService` impl

## Solution Proposal

- Add `RemoteMessage#getCounterPartyId()` 
- Adjust `TokenParameters`: 
  ```diff
  - private String audience;
  + private String counterPartyId;
  + private String counterPartyAddress;
  ```

### PR DESCRIPTION
**Pull request by wolf4ood**
## What this PR changes/adds

Improves the audience handling by adding an extension point `AudienceResolver` which each identity service or custom specific implementations should implement for mapping a `RemoteMessage` to an `aud` claim.

For doing that the `RemoteMessage` interface has been enriched with a `counterPartyId` field for those use case 
that needs the counter party id as `aud` claim.

Additionally the `counterPartyId` is added in the Catalog/Dataset API for supporting the audience mapping in the `Catalog` APIs

## Why it does that

iatp adotpion, refactoring

Closes #3747 

_Please be sure to take a look at the [contributing guidelines](https://github.com/eclipse-edc/.github/blob/main/CONTRIBUTING.md#submit-a-pull-request) and our [etiquette for pull requests](https://github.com/eclipse-edc/.github/blob/main/contributing/pr_etiquette.md)._

### CODE REVIEW

File: core/control-plane/transfer-core/src/main/java/org/eclipse/edc/connector/transfer/process/TransferProcessManagerImpl.java [hunk 1]
```diff
@@ -388,7 +387,7 @@ private void sendTransferStartMessage(TransferProcess process, DataFlowResponse
                 .dataAddress(dataFlowResponse.getDataAddress())
                 .policy(policy);
```
Comment (ndr-brt) [inline core/control-plane/transfer-core/src/main/java/org/eclipse/edc/connector/transfer/process/TransferProcessManagerImpl.java:388]: this line can be removed

File: core/control-plane/contract-core/src/main/java/org/eclipse/edc/connector/contract/policy/PolicyArchiveImpl.java [hunk 2]
```diff
@@ -32,8 +32,27 @@ public PolicyArchiveImpl(ContractNegotiationStore contractNegotiationStore) {
...
                 .map(contractNegotiationStore::findContractAgreement)
-                .map(ContractAgreement::getPolicy)
+                .map(this::mapAgreementPolicy)
                 .orElse(null);
     }
 
+    // TODO assignee and assigner should end up stored in the Agreement's policy as outlined here
+    //  https://github.com/International-Data-Spaces-Association/ids-specification/issues/195
+    //  As fallback we fill the assignee and the assigner from the consumer and provider id in
+    //  the contract agreement
```
Comment (ndr-brt) [inline core/control-plane/contract-core/src/main/java/org/eclipse/edc/connector/contract/policy/PolicyArchiveImpl.java:42]: should this deserve a different issue/PR? can be separated?

Comment (wolf4ood) [reply core/control-plane/contract-core/src/main/java/org/eclipse/edc/connector/contract/policy/PolicyArchiveImpl.java:42]: Yes i will raise an issue for supporting this

File: spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/callback/CallbackEventRemoteMessage.java [hunk 3]
```diff
@@ -56,6 +56,9 @@ public String getAuthCodeId() {
     public EventEnvelope<T> getEventEnvelope() {
         return envelope;
     }
-
-
+    
+    @Override
+    public String getCounterPartyId() {
+        return null;
```
[the remark this answers is not in the document]
Comment (wolf4ood) [reply spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/callback/CallbackEventRemoteMessage.java:62]: Yeah make sense to revisit this in future

**Review by paullatzelsperger — APPROVED** [review]

File: data-protocols/dsp/dsp-http-core/src/main/java/org/eclipse/edc/protocol/dsp/dispatcher/DspHttpRemoteMessageDispatcherImpl.java [hunk 4]
```diff
@@ -94,7 +99,7 @@ public <T, M extends RemoteMessage> CompletableFuture<StatusResult<T>> dispatch(
         }
 
         var tokenParameters = tokenParametersBuilder
-                .claims("aud", message.getCounterPartyAddress()) // enforce the audience, ignore anything a decorator might have set
+                .claims("aud", audienceResolver.resolve(message)) // enforce the audience, ignore anything a decorator might have set
```
[the remark this answers is not in the document]
Comment (wolf4ood) [reply data-protocols/dsp/dsp-http-core/src/main/java/org/eclipse/edc/protocol/dsp/dispatcher/DspHttpRemoteMessageDispatcherImpl.java]: Cannot use that, the deps is not there, added local constant

**Pull request merged**